src/flask_pyoidc/pyoidc_facade.py

```diff
@@ -216,6 +216,7 @@ def userinfo_request(self, access_token: str):
 
         return userinfo_response
 
+    @cachetools.cachedmethod(cache=lambda self: self._provider_configuration._cache)
```

Comment:
To improve security the caching needs to be done manually: the access token lifetime is returned in response as `exp` so the response should only be cached for `min(self.introspection_cache_configuration['max_ttl'], introspection_response['exp'])`, can this be done with `cachetools.TTLCache`?

Reply:
It caches the complete return output of the method. So the full token introspection response is cached otherwise the cached return output will break the caller method if I modify it.

Here's how the cache is being stored:

```python
TTLCache(
    {(<class 'cachetools.keys._HashedTuple'>, 'access_token', 'eyJhbG...'): <oic.extension.message.TokenIntrospectionResponse object at 0x000001ABE53603A0>}, maxsize=128, currsize=1)
```

Reply:
Yep, good point - that behaviour needs to stay the same.
Still the caching can't rely on decorator, since it only has one globally applied TTL - the TTL needs to be set per response (to respect the `exp`).
I haven't dug into the specifics of `cachetools`, so not sure if it's possible but I think that's a requirement for this change to be introduced.
(Otherwise I'd rather err on the side of caution and simply not cache the response.)

Reply:
There are 2 ways of achieving it:

1. In `OIDCAuthentication.introspect_token`, I can write a condition that checks if the `exp` of access token is less than the `ttl` of cache. If it is, then remove the cached token from the cache like this:

https://github.com/zamzterz/Flask-pyoidc/blob/50a67f31944868fbdf90dd2096fc3f82604cde14/src/flask_pyoidc/flask_pyoidc.py#L386-L393

```python
        if result.get('exp', 0) < time.time() + client._provider_configuration._cache.ttl:
            # Pop the cached token if it's about to expire.
            client._token_introspection_request.cache(client).pop(
                client.__token_introspection_request.cache_key(client, access_token=received_access_token))
```

The popping mechanism looks slightly complicated but it works. I'm already using it in `revoke_token` method.

2. Another way is to prevent it from caching in the first place by subclassing the `cachetools.TTLCache` and override the `__setitem__` method to include our condition.

Which one do you think will be more optimal?

Reply:
@zamzterz Oh I get it. You want short-lived access tokens to be cached but you want them to be removed from cache if they expire before the cache is expired to prevent unauthorized access. Nice design. I am working on point 2 to see if it can be achieved.

Reply:
@zamzterz Done. I hope you like it. 🙂

Now, the access token which is about to be cached will have its own TTL based on `exp`. If the access token will expire before the TTL set by the user, its cache will also be removed. See the new commit.